LEGO1/mxwavepresenter.cpp

```diff
@@ -159,63 +197,182 @@ void MxWavePresenter::StartingTickle()
 	}
 }
 
-// OFFSET: LEGO1 0x100b1ea0 STUB
+// OFFSET: LEGO1 0x100b1ea0
 void MxWavePresenter::StreamingTickle()
 {
-	// TODO
+	if (!m_currentChunk) {
+		if (!(m_action->GetFlags() & MxDSAction::Flag_Looping)) {
+			MxStreamChunk* chunk = FUN_100b5650();
+
+			if (chunk && chunk->GetFlags() & MxDSChunk::Flag_Bit2 && !(chunk->GetFlags() & MxDSChunk::Flag_Bit16)) {
+				chunk->SetFlags(chunk->GetFlags() | MxDSChunk::Flag_Bit8);
+
+				m_currentChunk = new MxStreamChunk;
+				MxU8* data = new MxU8[m_length];
+
+				// TODO
+
+				m_currentChunk->SetLength(m_length);
+				m_currentChunk->SetData(data);
+				m_currentChunk->SetTime(chunk->GetTime() + 1000);
+				m_currentChunk->SetFlags(MxDSChunk::Flag_Bit1);
+			}
+		}
+
+		MxMediaPresenter::StreamingTickle();
+	}
 }
 
-// OFFSET: LEGO1 0x100b20c0 STUB
+// OFFSET: LEGO1 0x100b20c0
 void MxWavePresenter::DoneTickle()
 {
-	// TODO
+	if (m_dsBuffer) {
+		DWORD dwCurrentPlayCursor, dwCurrentWriteCursor;
+		m_dsBuffer->GetCurrentPosition(&dwCurrentPlayCursor, &dwCurrentWriteCursor);
+
+		MxS8 result = dwCurrentPlayCursor / m_length;
+		if (m_action->GetFlags() & MxDSAction::Flag_Bit7 || m_action->GetFlags() & MxDSAction::Flag_Looping ||
+			m_unk64 != result || m_bytes + (m_length * result) <= dwCurrentPlayCursor)
+			MxMediaPresenter::DoneTickle();
+	}
+	else
+		MxMediaPresenter::DoneTickle();
 }
 
-// OFFSET: LEGO1 0x100b2130 STUB
+// OFFSET: LEGO1 0x100b2130
 void MxWavePresenter::AppendChunk(MxStreamChunk* p_chunk)
 {
-	// TODO
+	FUN_100b1bd0(p_chunk->GetData(), p_chunk->GetLength());
+	if (IsEnabled())
+		m_subscriber->FUN_100b8390(p_chunk);
 }
 
-// OFFSET: LEGO1 0x100b2160 STUB
+// OFFSET: LEGO1 0x100b2160
 undefined4 MxWavePresenter::PutData()
 {
-	// TODO
+	MxAutoLocker lock(&m_criticalSection);
+
+	if (IsEnabled()) {
+		switch (m_currentTickleState) {
+		case TickleState_Streaming:
+			if (m_currentChunk && FUN_100b1ba0()) {
+				FUN_100b1bd0(m_currentChunk->GetData(), m_currentChunk->GetLength());
+				m_subscriber->FUN_100b8390(m_currentChunk);
+				m_currentChunk = NULL;
+			}
+
+			if (!m_unk65) {
+				m_dsBuffer->SetCurrentPosition(0);
+
+				if (m_dsBuffer->Play(0, 0, DSBPLAY_LOOPING) == DS_OK)
+					m_unk65 = TRUE;
+			}
+			break;
+		case TickleState_Repeating:
+			if (m_unk65)
+				break;
+
+			m_dsBuffer->SetCurrentPosition(0);
+
+			if (m_dsBuffer->Play(0, 0, m_action->GetLoopCount() > 1) == DS_OK)
+				m_unk65 = TRUE;
```

Comment:
m_isPlaying, my only concern is in the MxWavePresenter::Enable function where it sets it to false

Reply:
Seems reasonable. I believe the reason it's set to false there is because it's meant to "reset", and it will be set to `TRUE` once `Play` has been called again.